bridges/prometheus/producer.go

```diff
@@ -0,0 +1,254 @@
+		}
+		otelGauge.DataPoints[i] = dp
+	}
+	return otelGauge
+}
+
+func convertCounter(metrics []*dto.Metric, now time.Time) metricdata.Sum[float64] {
+	otelCounter := metricdata.Sum[float64]{
+		DataPoints:  make([]metricdata.DataPoint[float64], len(metrics)),
+		Temporality: metricdata.CumulativeTemporality,
+		IsMonotonic: true,
+	}
+	for i, m := range metrics {
+		dp := metricdata.DataPoint[float64]{
+			Attributes: convertLabels(m.GetLabel()),
+			StartTime:  processStartTime,
+			Time:       now,
+			Value:      m.GetCounter().GetValue(),
+			Exemplars:  []metricdata.Exemplar[float64]{convertExemplar(m.GetCounter().GetExemplar())},
+		}
+		createdTs := m.GetCounter().GetCreatedTimestamp()
+		if createdTs.IsValid() {
+			dp.StartTime = createdTs.AsTime()
+		}
+		if m.GetTimestampMs() != 0 {
+			dp.Time = time.UnixMilli(m.GetTimestampMs())
+		}
+		otelCounter.DataPoints[i] = dp
+	}
+	return otelCounter
+}
+
+func convertHistogram(metrics []*dto.Metric, now time.Time) metricdata.Histogram[float64] {
+	// TODO: support converting Prometheus "native" histograms to OTel exponential histograms.
+	otelHistogram := metricdata.Histogram[float64]{
+		DataPoints:  make([]metricdata.HistogramDataPoint[float64], len(metrics)),
+		Temporality: metricdata.CumulativeTemporality,
+	}
+	for i, m := range metrics {
+		bounds, bucketCounts, exemplars := convertBuckets(m.GetHistogram().GetBucket())
+		dp := metricdata.HistogramDataPoint[float64]{
+			Attributes:   convertLabels(m.GetLabel()),
+			StartTime:    processStartTime,
+			Time:         now,
+			Count:        m.GetHistogram().GetSampleCount(),
+			Sum:          m.GetHistogram().GetSampleSum(),
+			Bounds:       bounds,
+			BucketCounts: bucketCounts,
+			Exemplars:    exemplars,
+		}
+		createdTs := m.GetHistogram().GetCreatedTimestamp()
+		if createdTs.IsValid() {
+			dp.StartTime = createdTs.AsTime()
+		}
+		if m.GetTimestampMs() != 0 {
+			dp.Time = time.UnixMilli(m.GetTimestampMs())
+		}
+		otelHistogram.DataPoints[i] = dp
+	}
+	return otelHistogram
+}
+
+func convertBuckets(buckets []*dto.Bucket) ([]float64, []uint64, []metricdata.Exemplar[float64]) {
+	bounds := make([]float64, len(buckets)-1)
+	bucketCounts := make([]uint64, len(buckets))
+	exemplars := make([]metricdata.Exemplar[float64], 0)
+	for i, bucket := range buckets {
+		// The last bound is the +Inf bound, which is implied in OTel, but is
+		// explicit in Prometheus. Skip the last boundary, and assume it is the
+		// +Inf bound.
+		if i < len(bounds) {
+			bounds[i] = bucket.GetUpperBound()
+		}
+		bucketCounts[i] = bucket.GetCumulativeCount()
+		if bucket.GetExemplar() != nil {
+			exemplars = append(exemplars, convertExemplar(bucket.GetExemplar()))
+		}
+	}
+	return bounds, bucketCounts, exemplars
+}
+
+func convertLabels(labels []*dto.LabelPair) attribute.Set {
+	kvs := make([]attribute.KeyValue, len(labels))
+	for i, l := range labels {
+		kvs[i] = attribute.String(l.GetName(), l.GetValue())
```

Comment:
Should we have a test case for this line?

Reply:
done